File: PDOO/p4/Civitas/src/civitas/TituloPropiedad.java

package civitas;


public class TituloPropiedad {
    private float alquilerBase;
    private static float factorInteresesHipoteca = 1.1f;
    private float factorRevalorizacion;
    private float hipotecaBase;
    private Boolean hipotecado;
    /*private*/ public String nombre;
    private int numCasas;
    private int numHoteles;
    private float precioCompra;
    private float precioEdificar;
    
    private Jugador propietario;
    
    void actualizaPropietarioPorConversion(Jugador jugador){
        propietario = jugador;
    }
    
    Boolean cancelarHipoteca(Jugador jugador){
        Boolean operacionRealizada = false;
        if (getHipotecado() && esEsteElPropietario(jugador)){
            propietario.paga(getImporteCancelarHipoteca());
            hipotecado = false;
            operacionRealizada = true;
        }
        return operacionRealizada;
    }
    
    int cantidadCasasHoteles(){
        return (getNumCasas() + getNumHoteles());
    }
    
    Boolean comprar(Jugador jugador){
        Boolean resultado = false;
        if (!tienePropietario()){
            actualizaPropietarioPorConversion(jugador);
            //propietario = jugador;
            propietario.paga(getPrecioCompra());
            resultado = true;
        }
        return resultado;
    }
    
    Boolean construirCasa(Jugador jugador){
        Boolean resultado = false;
        if (esEsteElPropietario(jugador)){
            propietario.paga(precioEdificar);
            numCasas++;
            resultado = true;
        }
        return resultado;
    }
    
    Boolean construirHotel(Jugador jugador){
        Boolean resultado = false;
        if (esEsteElPropietario(jugador)){
            propietario.paga(precioEdificar);
            numHoteles++;
            resultado = true;
        }
        return resultado;
    }
    
    Boolean derruirCasas(int n, Jugador jugador){
        Boolean operacionRealizada = false;
        if (esEsteElPropietario(jugador) && getNumCasas() >= n){
            numCasas -= n;
            operacionRealizada = true;
        }
        return operacionRealizada;
    }
    
    /*private*/ protected Boolean esEsteElPropietario(Jugador jugador){
        Boolean resultado = false;
        if (propietario == jugador){
            resultado = true;
        }
        return resultado;
    }
    
    public Boolean getHipotecado(){
        return hipotecado;
    }
    
    float getImporteCancelarHipoteca(){
        return (getImporteHipoteca() * factorInteresesHipoteca);
    }
    
    private float getImporteHipoteca(){
        float cantidadRecibida;
        cantidadRecibida = hipotecaBase * (1 + (numCasas * 0.5f) + (numHoteles * 2.5f));
        return cantidadRecibida;
    }
    
    String getNombre(){
        return nombre;
    }

    int getNumCasas(){
        return numCasas;
    }
    
    int getNumHoteles(){
        return numHoteles;
    }
    
    /*private*/ protected float getPrecioAlquiler(){
        float precioAlquiler;
        precioAlquiler = alquilerBase * (1 + (numCasas * 0.5f) + (numHoteles * 2.5f));
        if (getHipotecado() || propietarioEncarcelado()){
            precioAlquiler = 0;
        }
        return precioAlquiler;
    }
    
    float getPrecioCompra(){
        return precioCompra;
    }
    
    float getPrecioEdificar(){
        return precioEdificar;
    }
    
    private float getPrecioVenta(){
        float precioVenta;
        precioVenta = precioCompra + (numCasas + 5 * numHoteles) * precioEdificar * factorRevalorizacion;
        return precioVenta;
    }
    
    /*package*/ /*protected*/ Jugador getPropietario(){
        return propietario;
    }
    
    Boolean hipotecar(Jugador jugador){
        Boolean operacionRealizada = false;
        if (!hipotecado && esEsteElPropietario(jugador)){
            propietario.recibe(getImporteHipoteca());
            hipotecado = true;
            operacionRealizada = true;
        }
        return operacionRealizada;
    }
    
    private Boolean propietarioEncarcelado(){
        Boolean resultado = false;
        if (tienePropietario()){
            resultado = propietario.isEncarcelado();
        }
        return resultado;
    }
    
    Boolean tienePropietario(){
        Boolean resultado = false;
        if (propietario != null){
            resultado = true;
        }
        return resultado;
    }
    
    TituloPropiedad(String nom, float ab, float fr, float hb, float pc, float pe){
        nombre = nom;
        alquilerBase = ab;
        factorRevalorizacion = fr;
        hipotecaBase = hb;
        precioCompra = pc;
        precioEdificar = pe;
        hipotecado = false;
        numCasas = 0;
        numHoteles = 0;
    }
    
    @Override
    public String toString(){
        return "TituloPropiedad { nombre=" + nombre + ", numero de hoteles=" + numHoteles +
                ", numero de casas=" + numCasas + ", precio de compra=" + precioCompra + 
                ", precioEdificar=" + precioEdificar + ", hipotecada=" + hipotecado + 
                ", alquilerBase=" + alquilerBase + ", factorRevalorizacion=" + factorRevalorizacion + 
                ", hipotecaBase=" + hipotecaBase + " }";
    }
    
    void tramitarAlquiler(Jugador jugador){
        if (tienePropietario() && !esEsteElPropietario(jugador)){
            jugador.pagaAlquiler(getPrecioAlquiler());
            propietario.recibe(getPrecioAlquiler());
        }
    }
    
    Boolean vender(Jugador jugador){
        Boolean resultado = false;
        if (esEsteElPropietario(jugador)){
            propietario.recibe(getPrecioVenta());
            numCasas = 0;
            numHoteles = 0;
            propietario = null;
            resultado = true;
        }
        return resultado;
    }
}

File: PDOO/p4/civitas_rb/lib/titulo_propiedad.rb
require_relative 'jugador'

module Civitas

class Titulo_propiedad
  
  @@factorintereseshipoteca=1.1
  
  def initialize(nom, ab, fr, hb, pc, pe)
    @nombre=nom
    @alquilerbase=ab
    @factorrevalorizacion=fr
    @factorintereseshipoteca= 1.1
    @hipotecabase=hb
    @preciocompra=pc
    @precioedificar=pe
    @hipotecado=false
    @numcasas=0
    @numhoteles=0
    @propietario=nil
  end
  
  def actualizar_propietario_por_conversion(jugador)
    @propietario=jugador
  end
  
  def cancelar_hipoteca(jugador)
    operacionrealizada=false
    if @hipotecado and es_este_el_propietario(jugador)
      @propietario.paga(getimportecancelarhipoteca)
      @hipotecado=false
      operacionrealizada=true
    end
    return operacionrealizada
  end
  
  def cantidadcasashoteles
    return (@numcasas+@numhoteles)
  end
  
  def comprar(jugador)
    resultado=false
    if !tienepropietario
      actualizar_propietario_por_conversion(jugador)
      @propietario.paga(@preciocompra)
      resultado=true
    end
    return resultado
  end
  
  def construircasa(jugador)
    resultado=false
    if es_este_el_propietario(jugador)
      @propietario.paga(@precioedificar)
      @numcasas=@numcasas+1
      resultado=true
    end
    return resultado
  end
  
  def construirhotel(jugador)
    resultado=false
    if es_este_el_propietario(jugador)
      @propietario.paga(5 * @precioedificar)
      @numhoteles=@numhoteles+1
      resultado=true
    end
    return resultado
  end
  
  def derruircasas(n, jugador)
    operacionrealizada=false
    if es_este_el_propietario(jugador) and @numcasas>=n
      @numcasas=@numcasas-n
      operacionrealizada=true
    end
    return operacionrealizada
  end
  
  def es_este_el_propietario(jugador)
    resultado=false
    if @propietario==jugador
      resultado=true
    end
    return resultado
  end
  
  def getimportecancelarhipoteca
    return (getimportehipoteca * @factorintereseshipoteca)
  end
  
  def getimportehipoteca
    cantidadrecibida= (1+ (@numcasas * 0.5) + (@numhoteles * 2.5)) * @hipotecabase
    return cantidadrecibida
  end
    
  def getprecioalquiler
    precioalquiler = (1 + (@numcasas*0.5) + (@numhoteles*2.5)) * @alquilerbase
    if @hipotecado or propietarioencarcelado
      precioalquiler=0
    end
    return precioalquiler
  end
  
  def getprecioventa
    precioventa = @preciocompra + (@numcasas + 5 * @numhoteles) * @precioedificar * @factorrevalorizacion
    return precioventa
  end
  
  def hipotecar(jugador)
    operacionrealizada =false
    if !@hipotecado and es_este_el_propietario(jugador)
      @propietario.recibe(getimportehipoteca)
      @hipotecado=true
      operacionrealizada=true
    end
    return operacionrealizada
  end
  
  def propietarioencarcelado
    resultado=false
    if tienepropietario
      resultado=@propietario.encarcelado
    end
    return resultado
  end
  
  def tienepropietario
    resultado= false
    if @propietario != nil
      resultado=true
    end
    return resultado
  end
  
  def tostring
    return "TituloPropiedad { nombre=#{@nombre}, hipotecada=#{@hipotecado}, precio de compra=#{@preciocompra}, alquilerBase= #{@alquilerbase},
                  factorRevalorizacion=#{@factorrevalorizacion}, hipotecaBase=#{@hipotecabase}, precioEdificar= #{@precioedificar}, numero de hoteles=#{@numhoteles},
                  numero de casas=#{@numcasas} }"
  end
  
  def tramitaralquiler(jugador)
    if tienepropietario and !es_este_el_propietario(jugador)
      jugador.paga_alquiler(getprecioalquiler)
      @propietario.recibe(getprecioalquiler)
    end
  end
  
  def vender(jugador)
    resultado=false
    if es_este_el_propietario(jugador)
      @propietario.recibe(getprecioventa)
      @numcasas=0
      @numhoteles=0
      @propietario=nil
      resultado=true
    end
    return resultado
  end
  

  attr_accessor :hipotecado, :nombre, :numcasas, :numhoteles, :preciocompra, :precioedificar, :propietario
  
  private :es_este_el_propietario, :getimportehipoteca, :getprecioalquiler
end
end